les/serverpool.go

```diff
@@ -17,904 +17,354 @@
-}
-
-func (a *poolEntryAddress) Weight() int64 {
-	t := time.Duration(mclock.Now() - a.lastSeen)
-	return int64(1000000*math.Exp(-float64(t)/float64(discoverExpireConst)-float64(a.fails)*addrFailDropLn)) + 1
-}
-
-func (a *poolEntryAddress) strKey() string {
-	return a.ip.String() + ":" + strconv.Itoa(int(a.port))
-}
-
-// poolStats implement statistics for a certain quantity with a long term average
-// and a short term value which is adjusted exponentially with a factor of
-// pstatRecentAdjust with each update and also returned exponentially to the
-// average with the time constant pstatReturnToMeanTC
-type poolStats struct {
-	sum, weight, avg, recent float64
-	lastRecalc               mclock.AbsTime
-}
-
-// init initializes stats with a long term sum/update count pair retrieved from the database
-func (s *poolStats) init(sum, weight float64) {
-	s.sum = sum
-	s.weight = weight
-	var avg float64
-	if weight > 0 {
-		avg = s.sum / weight
+	close(s.quit)
+	s.persistClock()
+	s.ns.Stop()
+}
+
+// persistClock stores the persistent absolute time into the database
+func (s *serverPool) persistClock() {
+	var clockEnc [8]byte
+	binary.BigEndian.PutUint64(clockEnc[:], uint64(s.clock.Now()+s.clockOffset))
+	s.db.Put(s.dbClockKey, clockEnc[:])
+}
+
+// registerPeer implements serverPeerSubscriber
+func (s *serverPool) registerPeer(p *serverPeer) {
+	s.ns.SetState(p.Node(), sfConnected, sfDialed, 0)
+	nvt := s.vt.Register(p.ID())
+	s.ns.SetField(p.Node(), sfiConnectedStats, nvt.RtStats())
+	p.setValueTracker(s.vt, nvt)
+	p.updateVtParams()
+}
+
+// unregisterPeer implements serverPeerSubscriber
+func (s *serverPool) unregisterPeer(p *serverPeer) {
+	s.updateNode(p.Node(), true, true, dialCost)
+	s.ns.SetState(p.Node(), nodestate.Flags{}, sfConnected, 0)
+	s.ns.SetField(p.Node(), sfiConnectedStats, nil)
+	s.vt.Unregister(p.ID())
+	p.setValueTracker(nil, nil)
+}
+
+// getTimeout calculates the current recommended timeout. This value is used by
+// the client as a "soft timeout" value. It also affects the service value calculation
+// of individual nodes.
+func (s *serverPool) getTimeout() time.Duration {
+	now := s.clock.Now()
+	s.timeoutLock.RLock()
+	timeout := s.timeout
+	refreshed := s.timeoutRefreshed
+	s.timeoutLock.RUnlock()
+	if refreshed != 0 && time.Duration(now-refreshed) < timeoutRefresh {
+		return timeout
+	}
+	rts := s.vt.RtStats()
+	rts.Add(time.Second*2, 10, s.vt.StatsExpFactor())
+	timeout = minTimeout
+	if t := rts.Timeout(0.1); t > timeout {
+		timeout = t
+	}
+	if t := rts.Timeout(0.5) * 2; t > timeout {
+		timeout = t
+	}
+	s.timeoutLock.Lock()
+	if s.timeout != timeout {
+		s.timeout = timeout
+		s.timeWeights = lpc.TimeoutWeights(timeout)
+	}
+	s.timeoutRefreshed = now
+	s.timeoutLock.Unlock()
+	return timeout
+}
+
+// updateNode calculates the selection weight and the proposed redial wait time of the given node
+func (s *serverPool) updateNode(node *enode.Node, calculateSessionValue, redialWait bool, addDialCost int64) {
```

Comment:
I would recommend to split this big function into several parts.

For example `recalSessionValue`, `recalRedialInterval`, etc. It's not easy to determine the functionality of this function, with lots of paramters.